### DISCUSSION TITLE
How to retrieve all activities for the current user?

### DISCUSSION THREAD

**Discussion by nickywan123** [spatie/laravel-activitylog#946] [Q&A] [answered]
I followed the docs and implemented this package to my app. However, let say I have a `Transaction ` model where it also logs into the `activity_log` table. There's no user ID on the table but just the `subject_id` that belongs to `transactions` table. My goal is to display all the activities that belong or done by the user so I can show it in my view. 

I don't see any way to retrieve it on the docs, do I need to customize the table and add a `user_id`?

**Comment by Gummibeer** [comment, marked as answer]
There's no `user_id` but a `causer` which is for the user or any other entity doing activities.

https://spatie.be/docs/laravel-activitylog/v4/basic-usage/logging-activity#setting-a-causer
https://spatie.be/docs/laravel-activitylog/v4/advanced-usage/logging-model-events#using-the-causesactivity-trait
https://spatie.be/docs/laravel-activitylog/v4/advanced-usage/define-causer-for-runtime
https://spatie.be/docs/laravel-activitylog/v4/api/causer-resolver

**Reply by nickywan123** [reply]
hi, what if I have a callback from a payment gateway and save the callback data in `Transaction` table where I cached the user_id to it. Does the activity log table grabs the data from the `Transaction` table "right" after it is saved inside it?

So in my `TapActivity()` method in the model, I can do something like `$activity->causer_id = $this->user_id`?

**Reply by Gummibeer** [reply]
This wouldn't be enough. You should use the polymorphic methods to define the relationship.
If I remember right it should be `$activity->causer()->associate($this->user)`
https://github.com/spatie/laravel-activitylog/blob/1d8acdac3e1f3d6937f8d10fd005de71b0350bc4/src/ActivityLogger.php#L67

**Reply by nickywan123** [reply]
I wrote it like this:
```
 public function tapActivity(Activity $activity, string $eventName)
    {   
        $activity->causer_type = "App\User";
        $activity->causer_id = $this->id;
    }

```
and it does works as I tested. Alternatively, if user is logged, you can also do it like this:

```
 public function tapActivity(Activity $activity, string $eventName)
    {   
     
        $activity->causer_id = auth()->id();
    }
```
and it will populate `causer_type` as well as  `causer_id` automatically.

**Reply by Gummibeer** [reply]
If you are logged in you don't have to do anything as we associate the logged in user by default to all activities.
But assigning a polymorphic relationship manually has two big problems.
1) refactoring the class doesn't change the type automatically - moving or renaming
2) it doesn't use the morph map to customize the type - using a morph map is highly recommended to decouple your database from code/class structure.

**Reply by nickywan123** [reply]
So how should I refactor this to : 

```
 public function tapActivity(Activity $activity, string $eventName)
    {   
        $activity->causer_type = "App\User";
        $activity->causer_id = $this->user_id;
    }
```
?

My current way of doing works, is it okay?

The reason I do it this way, for example if a user successfully register an account, it does not logged the causer details, that's why I use ```tapActivity```. Another reason is because when I am in the payment processor page(external site), it does not logged the user in my system so after successful transaction, it brings user back to my system where it couldn't get the logged user.

**Reply by Gummibeer** [reply]
Like I already wrote. Nothing wrong about the tap method. But you shouldn't assign the polymorphic attributes manually but let eloquent do it.

> This wouldn't be enough. You should use the polymorphic methods to define the relationship.
> If I remember right it should be `$activity->causer()->associate($this->user)`
> https://github.com/spatie/laravel-activitylog/blob/1d8acdac3e1f3d6937f8d10fd005de71b0350bc4/src/ActivityLogger.php#L67
